pep-0634.rst

```diff
@@ -0,0 +1,698 @@
+
+A fixed-length sequence pattern matches the subpatterns to
+corresponding items of the subject sequence, from left to right.
+Matching stops (with a failure) as soon as a subpattern fails.  If all
+subpatterns succeed in matching their corresponding item, the sequence
+pattern succeeds.
+
+TODO: Describe variable-length sequence patterns.  (Brandt?)
+
+
+.. _mapping_pattern:
+
+Mapping Patterns
+~~~~~~~~~~~~~~~~
+
+Syntax::
+
+  mapping_pattern: '{' items_pattern? '}'
+  items_pattern: ','.key_value_pattern+ ','?
+  key_value_pattern:
+      | (literal_pattern | constant_pattern) ':' or_pattern
+      | '**' NAME
+
+(Note that ``'**' "_"`` is disallowed by this grammar.)
+
+A mapping pattern fails if the subject value is not an instance of
+``collections.abc.Mapping``.
+
+A mapping pattern succeeds if every key given in the mapping pattern
+matches the corresponding item of the subject mapping.  If a ``'**'
+NAME`` form is present, that name is bound to a ``dict`` containing
+remaining key-value pairs from the subject mapping.
+
+If duplicate keys are detected in the mapping pattern, the pattern is
+considered invalid, and a ``ValueError`` is raised.
+
+Key-value pairs are matched using the two-argument form of the
+subject's ``get()`` method.  As a consequence, matched key-value pairs
+must already be present in the mapping, and not created on-the-fly by
+``__missing__`` or ``__getitem__``.  For example,
+``collections.defaultdict`` instances will only be matched by patterns
+with keys that were already present when the ``match`` block was
+entered.
+
+
+.. _class_pattern:
+
+Class Patterns
+~~~~~~~~~~~~~~
+
+Simplified syntax::
+
+  class_pattern:
+      | name_or_attr '(' ')'
+      | name_or_attr '(' ','.pattern+ ','? ')'
+      | name_or_attr '(' ','.keyword_pattern+ ','? ')'
+      | name_or_attr '(' ','.pattern+ ',' ','.keyword_pattern+ ','? ')'
+  keyword_pattern: NAME '=' or_pattern
+
+
+A class pattern provides support for destructuring arbitrary objects.
+There are two possible ways of matching on object attributes: by position
+like ``Point(1, 2)``, and by name like ``Point(x=1, y=2)``. These
+two can be combined, but a positional match cannot follow a match by name.
+Each item in a class pattern can be an arbitrary pattern. A simple
+example::
+
+  match shape:
+      case Point(x, y):
+          ...
+      case Rectangle(x0, y0, x1, y1, painted=True):
+          ...
+
+Whether a match succeeds or not is determined by the equivalent of an
+``isinstance`` call.  If the subject (``shape``, in the example) is not
+an instance of the named class (``Point`` or ``Rectangle``), the match
+fails.  Otherwise, it continues (see details in the `runtime`_
+section).
+
+The named class must inherit from ``type``.  It may be a single name
+or a dotted name (e.g. ``some_mod.SomeClass`` or ``mod.pkg.Class``).
+The leading name must not be ``_``, so e.g. ``_(...)`` and
+``_.C(...)`` are invalid. Use ``object(foo=_)`` to check whether the
```

Comment:
Is the restriction on `_(...)` and `_.C(...)` still true? The current implementation doesn't raise here, and I sort of remember walking this back. Not sure though.

Reply:
It looks like it no longer applies: https://github.com/gvanrossum/patma/issues/73#issuecomment-648832371

I'll go ahead and remove this sentence.

Reply:
Let's walk it back. I'm out of time, whoever gets to it first can do it.